history-network.md

```diff
@@ -0,0 +1,300 @@
+
+NOTE: The `Offer`/`Accept`/`Store` messages do not conform to the request/response paradigm of Discovery v5. We plan to propose a change to the base protocol definition to loosen the language.
+
+#### Ping (0x01)
+
+Communicate `radius` information of the sender, and request `radius` information of the recipient.
+
+```
+protocol-message-type = 0x01
+protocol-message      = Container(radius: uint256)
+```
+
+#### Pong (0x02)
+
+In response to a `Ping` message, communicate `radius` information of the sender.
+
+```
+protocol-message-type = 0x02
+protocol-message      = Container(radius: uint256)
+```
+
+#### FindNode (0x03)
+
+Request a list of ENR values for nodes whose `logdistance` from the recipient's `node-id` is equal to one of the specified values.
+
+```
+protocol-message-type = 0x03
+protocol-message      = Container(distances: List[uint8, 256])
+```
+
+Each element of `distances` **MUST** be unique. A zero value in the `distances` list represents a request for the recipient's ENR value.
+
+#### FoundNodes (0x04)
+
+In response to a `FindNode` message, communicate a list of RLP-encoded ENR values for nodes whose `logdistance` value satisfies the request.
+
+```
+protocol-message-type = 0x04
+protocol-message      = Container(total: uint8, enrs: List[Bytes, 32])
+```
+
+The recipient may know of more than 32 nodes that satisfy the request. In this case, the recipient will need to send multiple `FoundNodes` messages in response to the `FindNode` message.
+
+Here, `total` denotes the total number of `FoundNodes` messages that the sender of the `FindNode` message should expect in response.
+
+Each ENR **MUST** be unique to some `node-id`. Each ENR **MUST** correspond to an element in the `distances` field of the `FindNode` request.
+
+#### FindContent (0x05)
+
+Request the data for `content` with `content-key`, or a list of ENR values for nodes whose `distance` from `content-id` is smaller than that of the recipient.
+
+The encoding of `content-key` for all possible data items is specified above.
+
+```
+protocol-message-type = 0x05
+protocol-message      = Container(content-key: Bytes)
+```
+
+#### FoundContent (0x06)
+
+In response to a `FindContent` message, communicate one of the following:
+
+* A connection ID for a uTP stream to transmit the requested  data
+* A list of RLP-encoded ENR values with maximum length of 32
+* A byte-array that encodes the requested data
+
+```
+protocol-message-type = 0x06
+protocol-message      = Container(connection-id: Bytes4, enrs: List[Bytes, 32], content: Bytes)
+```
+
+If the node does not hold the requested content, and the node does not know of any nodes with eligible ENR values, then the node should return `connection-id` as a zero byte array, `enrs` as an empty list and `content` as an empty byte-array.
```

Comment:
It would be nice to specify also what should happen when node receives malformed content key i.e from wrong chain or with wrong content type .